File: console_human_player.rb
require_relative "player"

class ConsoleHumanPlayer < Player

  def initialize(goban, color)
    super(true, goban)
    set_color(color)
    @debug_ai = nil
  end
  
  def get_move(color=@color)
    @goban.console_display
    puts "What is #{Grid::COLOR_NAMES[color]}/#{Grid::COLOR_CHARS[color]} move? (or 'help')"
    return get_answer
  end

  def attach_debug_ai(ai)
    @debug_ai = ai
  end

  def get_ai_eval(i,j)
    if @debug_ai
      @debug_ai.prepare_eval
      score = @debug_ai.eval_move(i,j)
      $log.debug("==> AI would rank this move (#{i},#{j}) as #{score}")
    end
  end

  def propose_score
    puts "Do you accept this score? (y/n)"
    return get_answer(["y","n"]) == "y"
  end

private

  def get_answer(valid_ones=nil)
    while true do
      answer = gets.downcase.strip
      next if answer == ""
      if valid_ones and ! valid_ones.find_index(answer)
        puts "Valid answers: "+valid_ones.join(",")
        next
      end
      return answer
    end
  end
  
end

File: js/ConsoleHumanPlayer.js
//Translated from console_human_player.rb using babyruby2js
'use strict';

var inherits = require('util').inherits;
var Grid = require('./Grid');
var main = require('./main');
var Player = require('./Player');

/** @class */
function ConsoleHumanPlayer(goban, color) {
    Player.call(this, true, goban);
    setColor(color);
    this.debugAi = null;
}
inherits(ConsoleHumanPlayer, Player);
module.exports = ConsoleHumanPlayer;

ConsoleHumanPlayer.prototype.getMove = function (color) {
    if (color === undefined) color = this.color;
    this.goban.consoleDisplay();
    console.log('What is ' + Grid.COLOR_NAMES[color] + '/' + Grid.COLOR_CHARS[color] + ' move? (or \'help\')');
    return this.getAnswer();
};

ConsoleHumanPlayer.prototype.attachDebugAi = function (ai) {
    this.debugAi = ai;
};

ConsoleHumanPlayer.prototype.getAiEval = function (i, j) {
    if (this.debugAi) {
        this.debugAi.prepareEval();
        var score = this.debugAi.evalMove(i, j);
        return main.log.debug('==> AI would rank this move (' + i + ',' + j + ') as ' + score);
    }
};

ConsoleHumanPlayer.prototype.proposeScore = function () {
    console.log('Do you accept this score? (y/n)');
    return this.getAnswer(['y', 'n']) === 'y';
};

//private;
ConsoleHumanPlayer.prototype.getAnswer = function (validOnes) {
    if (validOnes === undefined) validOnes = null;
    while (true) {
        var answer = gets().toLowerCase().trim();
        if (answer === '') {
            continue;
        }
        if (validOnes && !validOnes.findIndex(answer)) {
            console.log('Valid answers: ' + validOnes.join(','));
            continue;
        }
        return answer;
    }
};

// E02: unknown method gets()
// E02: unknown method find_index(...)